api/src/auth/drivers/saml.ts

```diff
@@ -0,0 +1,172 @@
+import { getAuthProvider } from '../../auth';
+import { COOKIE_OPTIONS } from '../../constants';
+import env from '../../env';
+import { RecordNotUniqueException } from '../../exceptions/database/record-not-unique';
+import { InvalidCredentialsException, InvalidProviderException } from '../../exceptions';
+import { respond } from '../../middleware/respond';
+import { AuthenticationService, UsersService } from '../../services';
+import { AuthData, AuthDriverOptions, User } from '../../types';
+import asyncHandler from '../../utils/async-handler';
+import { getConfigFromEnv } from '../../utils/get-config-from-env';
+import { LocalAuthDriver } from './local';
+import { Url } from '../../utils/url';
+
+// tell samlify to use validator...
+samlify.setSchemaValidator(validator);
+
+export class SAMLAuthDriver extends LocalAuthDriver {
+	idp: any;
+	sp: any;
+	redirectUrl: string;
+	usersService: UsersService;
+	config: Record<string, any>;
+
+	constructor(options: AuthDriverOptions, config: Record<string, any>) {
+		super(options, config);
+
+		this.config = config;
+		this.usersService = new UsersService({ knex: this.knex, schema: this.schema });
+
+		const redirectUrl = new Url(env.PUBLIC_URL).addPath('auth', 'login', config.provider, 'callback');
+		this.redirectUrl = redirectUrl.toString();
+
+		this.sp = samlify.ServiceProvider(getConfigFromEnv(`AUTH_${config.provider.toUpperCase()}_SP`));
+		this.idp = samlify.IdentityProvider(getConfigFromEnv(`AUTH_${config.provider.toUpperCase()}_IDP`));
+	}
+
+	async fetchUserID(identifier: string) {
+		const user = await this.knex
+			.select('id')
+			.from('directus_users')
+			.whereRaw('LOWER(??) = ?', ['external_identifier', identifier.toLowerCase()])
+			.first();
+
+		return user?.id;
+	}
+
+	async getUserID(payload: Record<string, any>) {
+		const { provider, emailKey, identifierKey, nameKey, givenNameKey, familyNameKey, allowPublicRegistration } =
+			this.config;
+
+		const email = payload[emailKey ?? 'http://schemas.xmlsoap.org/ws/2005/05/identity/claims/emailaddress'];
+		const identifier = payload[identifierKey ?? 'http://schemas.xmlsoap.org/ws/2005/05/identity/claims/nameidentifier'];
+		const userID = await this.fetchUserID(identifier);
+
+		const userInfo = {
+			name: payload[nameKey ?? 'http://schemas.xmlsoap.org/ws/2005/05/identity/claims/name'],
+			given_name: payload[givenNameKey ?? 'http://schemas.xmlsoap.org/ws/2005/05/identity/claims/givenname'],
+			family_name: payload[familyNameKey ?? 'http://schemas.xmlsoap.org/ws/2005/05/identity/claims/surname'],
+			email,
+		};
+
+		if (userID) return userID;
+
+		// Is public registration allowed?
+		if (!allowPublicRegistration) {
+			logger.trace(`[SAML] User doesn't exist, and public registration not allowed for provider "${provider}"`);
+			throw new InvalidCredentialsException();
+		}
+
+		try {
+			await this.usersService.createOne({
+				provider,
+				first_name: userInfo.given_name,
+				last_name: userInfo.family_name,
```

Comment:
**suggestion (non-blocking):** I think this is more clear. Creating an object implies intent, but here we create the object only to use the `given_name` and `family_name` once. That being the case they might as well be variables more closely coupled with where they're used.

```suggestion
		if (userID) return userID;

		// Is public registration allowed?
		if (!allowPublicRegistration) {
			logger.trace(`[SAML] User doesn't exist, and public registration not allowed for provider "${provider}"`);
			throw new InvalidCredentialsException();
		}

		const firstName = payload[givenNameKey ?? 'http://schemas.xmlsoap.org/ws/2005/05/identity/claims/givenname'];
		const lastName = payload[familyNameKey ?? 'http://schemas.xmlsoap.org/ws/2005/05/identity/claims/surname'];

		try {
			await this.usersService.createOne({
				provider,
				first_name: firstName,
				last_name: lastName,
```